packages/peregrine/lib/talons/CheckoutPage/AddressBook/useAddressCard.js

```diff
@@ -0,0 +1,61 @@
+import { useCallback, useEffect, useMemo, useState, useRef } from 'react';
+
+export const useAddressCard = props => {
+    const { address, onEdit, onSelection } = props;
+    const { id: addressId } = address;
+
+    const [hasUpdate, setHasUpdate] = useState(false);
+    const hasRendered = useRef(false);
+
+    useEffect(() => {
+        if (address !== undefined) {
+            if (hasRendered.current) {
+                setHasUpdate(true);
+                setTimeout(() => {
+                    setHasUpdate(false);
+                }, 2000);
+            } else {
+                hasRendered.current = true;
+            }
+        }
+    }, [hasRendered, address]);
```

Comment:
This effect, and the one in `useShippingInformation` are a bit concerning. Using a side effect to check for updates is something that would normally indicate we need better communication between certain components.

Can you elaborate on how this works? We can pair review it if you'd prefer.

In the meantime, whenever you're setting a timeout in a side effect, you should [capture its `timeoutId`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout) and [return a function](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect) from the effect that [clears the timeout](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/clearTimeout). It may take a bit of finesse to get the lifecycle right.

Reply:
This was my solution to:
>If we can add an animation to the card when it receives data updates, that would be provide visual feedback on the change - both within the address view and on the checkout page

This seemed like a common solution to replicate behavior of `componentDidUpdate`, where I only want to fire the state update (`hasUpdate`) after the initial render of data. Using this state I attach a class to an element that triggers the animation, then with some time buffer, we remove the animation class just in case it will get triggered again with another update.

![](https://gyazo.com/5b306e89e7ce1ee64d747413bf7e2c7f.gif)

We do need to handle the unmount in case the user navigate away, so I will get that added. Let me know if you have some concerns about this approach; I'm sure a solution with callbacks drilled down into the forms would also work, but this seemed just as clean reacting to the state change of data.

Reply:
Cool. Ultimately I think this approach will work. 👍

Reply:
Added unmount cleanup in b5aec84.